src/libmongoc/src/mongoc/mongoc-collection.c

```diff
@@ -796,6 +796,59 @@ mongoc_collection_count_with_opts (
    RETURN (ret);
 }
 
+/* --------------------------------------------------------------------------
+ *
+ * _make_aggregate_for_edc --
+ *
+ *       Construct an aggregate pipeline with the following form:
+ *
+ *
+ *       { pipeline: [
+ *           { $collStats: { count: {} } },
+ *           { $group: { _id: 1, n: { $sum: $count } } },
+ *         ]
+ *       }
+ *
+ *--------------------------------------------------------------------------
+ */
+static void
+_make_aggregate_for_edc (const mongoc_collection_t *coll, bson_t *out)
+{
+   bson_t pipeline;
+   bson_t coll_stats_stage;
+   bson_t coll_stats_stage_doc;
+   bson_t group_stage;
+   bson_t group_stage_doc;
+   bson_t sum;
+   bson_t cursor_empty;
+   bson_t empty;
+   const char *keys[] = {"0", "1"};
+   int key = 0;
+
+   bson_init (out);
+   bson_append_utf8 (
+      out, "aggregate", 9, coll->collection, coll->collectionlen);
+   bson_append_document_begin (out, "cursor", 6, &cursor_empty);
+   bson_append_document_end (out, &cursor_empty);
+   bson_append_array_begin (out, "pipeline", 8, &pipeline);
+
+   bson_append_document_begin (&pipeline, keys[key++], 1, &coll_stats_stage);
+   bson_append_document_begin (&coll_stats_stage, "$collStats", 10, &coll_stats_stage_doc);
+   bson_append_document_begin (&coll_stats_stage_doc, "count", 5, &empty);
+   bson_append_document_end (&coll_stats_stage_doc, &empty);
+   bson_append_document_end (&coll_stats_stage, &coll_stats_stage_doc);
+   bson_append_document_end (&pipeline, &coll_stats_stage);
+
+   bson_append_document_begin (&pipeline, keys[key], 1, &group_stage);
+   bson_append_document_begin (&group_stage, "$group", 6, &group_stage_doc);
+   bson_append_int32 (&group_stage_doc, "_id", 3, 1);
+   bson_append_document_begin (&group_stage_doc, "n", 1, &sum);
+   bson_append_utf8 (&sum, "$sum", 4, "$count", 6);
+   bson_append_document_end (&group_stage_doc, &sum);
+   bson_append_document_end (&group_stage, &group_stage_doc);
+   bson_append_document_end (&pipeline, &group_stage);
+   bson_append_array_end (out, &pipeline);
```

Comment:
Using the `BSON_APPEND_` macros would make this a little easier to read as it removes the need to pass string lengths for everything. Compare `bson_append_utf8` for example:
```
bson_append_utf8 (out, "aggregate", 9, coll->collection, coll->collectionlen);

BSON_APPEND_UTF8 (out, "aggregate", coll->collection);
```

Depending on whether clang-format allows you to do so,  leveraging indentation could make this a little more readable as well:
```
BSON_APPEND_UTF8 (out, "aggregate", coll->collection);

[...]

BSON_APPEND_DOCUMENT_BEGIN (&pipeline, keys[key++], &coll_stats_stage);
   BSON_APPEND_DOCUMENT_BEGIN (&coll_stats_stage, "$collStats", &coll_stats_stage_doc);
      BSON_APPEND_DOCUMENT_BEGIN (&coll_stats_stage_doc, "count", &empty);
      bson_append_document_end (&coll_stats_stage_doc, &empty);
   bson_append_document_end (&coll_stats_stage, &coll_stats_stage_doc);
bson_append_document_end (&pipeline, &coll_stats_stage);
```

This could help make sense of the nesting a little better, but as I said, not sure if clang-format will allow you to do this.

Reply:
Awesome! I wasn't aware of these macros. I've added them in but unfortunately cannot do the indentation you described. As you predicted, clang-format will not allow indentation of regular statements like this.